File: mapping.py
"""Department grouping: raw Department/Area → (level1, level2_or_None).

The mapping source lives in department_mapping.csv so it can be edited
without touching code. Readme §department grouping is the spec.

Normalisation:
- Lookup is case-insensitive and whitespace-tolerant (strip + collapse
  internal whitespace), so "Pre-Admission " and "pre-admission" both hit
  the same row.
- Level display strings collapse by lowercase key but preserve the
  first-seen canonical form, so "Commercial: Facilities" and
  "Commercial: facilities" group together under whichever spelling
  appeared first in the CSV.
- Split is on the FIRST ":" only. A value like
  "Commercial: Facilities: Car park" becomes level1="Commercial",
  level2="Facilities: Car park".
"""

from __future__ import annotations

import csv
import re
from pathlib import Path


Mapping = dict[str, tuple[str, str | None]]


def _collapse_ws(s: str) -> str:
    return re.sub(r"\s+", " ", s.strip())


def _norm_key(s: str | None) -> str:
    return _collapse_ws(str(s)).lower() if s is not None else ""


def load_mapping(path: str | Path) -> Mapping:
    rows: list[tuple[str, str, str | None]] = []
    with open(path, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            orig = row.get("original")
            new = row.get("new")
            if not orig or not new:
                continue
            parts = new.split(":", 1)
            l1 = _collapse_ws(parts[0])
            l2 = _collapse_ws(parts[1]) if len(parts) > 1 else None
            if not l1:
                continue
            if l2 == "":
                l2 = None
            rows.append((orig.strip(), l1, l2))

    # first-seen canonical display per lowercase key
    canon_l1: dict[str, str] = {}
    canon_l2: dict[tuple[str, str], str] = {}
    for _, l1, l2 in rows:
        canon_l1.setdefault(l1.lower(), l1)
        if l2 is not None:
            canon_l2.setdefault((l1.lower(), l2.lower()), l2)

    mapping: Mapping = {}
    for orig, l1, l2 in rows:
        l1c = canon_l1[l1.lower()]
        l2c = canon_l2[(l1.lower(), l2.lower())] if l2 is not None else None
        mapping[_norm_key(orig)] = (l1c, l2c)
    return mapping


def map_department(raw: object, mapping: Mapping) -> tuple[str | None, str | None]:
    if raw is None:
        return (None, None)
    return mapping.get(_norm_key(str(raw)), (None, None))

File: app/src-tauri/src/mapping.rs
//! Department grouping — direct port of Python `mapping.py`.
//!
//! The CSV is baked in at compile time so there's nothing to install at
//! runtime. A future iteration can expose an override path via Settings.

use std::collections::HashMap;

const EMBEDDED_CSV: &str = include_str!("../resources/department_mapping.csv");

#[derive(Debug, Clone)]
pub struct DepartmentMapping {
    /// Key is `normalise_key(raw)` — case-insensitive, collapsed whitespace.
    lookup: HashMap<String, (String, Option<String>)>,
}

impl DepartmentMapping {
    pub fn load_embedded() -> Result<Self, String> {
        Self::from_csv_str(EMBEDDED_CSV)
    }

    pub fn from_csv_str(csv_text: &str) -> Result<Self, String> {
        let mut reader = csv::Reader::from_reader(csv_text.as_bytes());
        let mut rows: Vec<(String, String, Option<String>)> = Vec::new();

        for record in reader.deserialize::<Row>() {
            let Row { original, new } = record.map_err(|e| format!("bad mapping row: {e}"))?;
            let orig = original.trim().to_string();
            if orig.is_empty() || new.trim().is_empty() {
                continue;
            }
            let (l1, l2) = split_once_colon(&new);
            let l1 = collapse_ws(&l1);
            let l2 = l2.map(|s| collapse_ws(&s)).filter(|s| !s.is_empty());
            if l1.is_empty() {
                continue;
            }
            rows.push((orig, l1, l2));
        }

        // First-seen canonical display per lowercase key.
        let mut canon_l1: HashMap<String, String> = HashMap::new();
        let mut canon_l2: HashMap<(String, String), String> = HashMap::new();
        for (_, l1, l2) in &rows {
            canon_l1.entry(l1.to_lowercase()).or_insert_with(|| l1.clone());
            if let Some(l2v) = l2 {
                canon_l2
                    .entry((l1.to_lowercase(), l2v.to_lowercase()))
                    .or_insert_with(|| l2v.clone());
            }
        }

        let mut lookup: HashMap<String, (String, Option<String>)> = HashMap::new();
        for (orig, l1, l2) in rows {
            let l1c = canon_l1
                .get(&l1.to_lowercase())
                .cloned()
                .unwrap_or_else(|| l1.clone());
            let l2c = l2
                .as_ref()
                .and_then(|l2v| canon_l2.get(&(l1.to_lowercase(), l2v.to_lowercase())).cloned());
            lookup.insert(normalise_key(&orig), (l1c, l2c));
        }

        Ok(Self { lookup })
    }

    /// Returns `(level1, level2_or_None)` if mapped, else None.
    pub fn map(&self, raw: &str) -> Option<(String, Option<String>)> {
        self.lookup.get(&normalise_key(raw)).cloned()
    }
}

#[derive(serde::Deserialize)]
struct Row {
    original: String,
    new: String,
}

fn split_once_colon(s: &str) -> (String, Option<String>) {
    match s.split_once(':') {
        Some((l, r)) => (l.to_string(), Some(r.to_string())),
        None => (s.to_string(), None),
    }
}

fn collapse_ws(s: &str) -> String {
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}

fn normalise_key(s: &str) -> String {
    collapse_ws(s).to_lowercase()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_embedded_mapping() {
        let m = DepartmentMapping::load_embedded().unwrap();
        assert_eq!(
            m.map("CSSD"),
            Some(("Theatre".to_string(), Some("CSSD".to_string())))
        );
        // Case-insensitive + collapse whitespace
        assert_eq!(
            m.map("  cssd  "),
            Some(("Theatre".to_string(), Some("CSSD".to_string())))
        );
        // Trailing space in original ("Pre-Admission ") still matches
        assert_eq!(
            m.map("Pre-Admission"),
            Some((
                "Patient services".to_string(),
                Some("Pre-Admission".to_string())
            ))
        );
        // Level-1-only (Administration)
        assert_eq!(m.map("Administration"), Some(("Administration".to_string(), None)));
        assert!(m.map("Not A Dept").is_none());
    }
}
